cmd/geth/import_from_firehose.go

```diff
@@ -0,0 +1,306 @@
+		if head != nil {
+			startBlock = int(head.Number.Uint64() + 1)
+			fmt.Printf("No start block specified. Resuming from last imported block: %d\n", startBlock)
+		} else {
+			startBlock = 0 // start from genesis
+			fmt.Println("No start block specified and no local chain found. Starting from genesis (block 0).")
+		}
+	}
+
+	var totalBlocks int
+	currentBlock := startBlock
+	err := processFirehoseBlocksWithReconnect(endpoint, apiToken, &currentBlock, endBlock, batchSize, workerCount, bufferSize, chainID, externalRpc, func(blocks []*types.Block, batchNum int) error {
+		if len(blocks) == 0 {
+			return nil
+		}
+		firstNum := blocks[0].NumberU64()
+		lastNum := blocks[len(blocks)-1].NumberU64()
+		if _, err := chain.InsertChain(blocks); err != nil {
+			fmt.Printf("failed to import batch %d (blocks %d-%d): %v\n", batchNum, firstNum, lastNum, err)
+			return err
+		}
+		fmt.Printf("Imported batch %d of %d blocks (blocks %d-%d)\n", batchNum, len(blocks), firstNum, lastNum)
+		totalBlocks += len(blocks)
+		currentBlock = int(lastNum + 1)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Import completed successfully. Total blocks imported: %d\n", totalBlocks)
+	return nil
+}
+
+func processFirehoseBlocksWithReconnect(
+	endpoint string,
+	apiToken string,
+	startBlock *int,
+	endBlock uint64,
+	batchSize int,
+	workerCount int,
+	bufferSize int,
+	chainID *big.Int,
+	externalRpc string,
+	handler func(blocks []*types.Block, batchNum int) error,
+) error {
+	maxRetries := 100
+	attempt := 0
+
+	for attempt < maxRetries {
+		prev := *startBlock
+
+		err := retry.Do(
+			func() error {
+				return processFirehoseBlocks(endpoint, apiToken, startBlock, endBlock, batchSize, workerCount, bufferSize, chainID, externalRpc, handler)
+			},
+			retry.Attempts(1),
+			retry.DelayType(retry.BackOffDelay),
+			retry.Delay(time.Second*2),
+			retry.MaxDelay(time.Minute*5),
+		)
+
+		if err == nil {
+			return nil // success
+		}
+
+		if *startBlock > prev {
+			fmt.Printf("Progress detected (%d → %d), resetting attempts\n", prev, *startBlock)
+			attempt = 0
+			continue
+		}
+
+		if isRetryableError(err) {
+			fmt.Printf("Retryable error: %v\n", err)
+			attempt++
+			continue
+		}
+
+		// Non-retryable error
+		return err
+	}
+
+	return fmt.Errorf("exceeded max retries (%d)", maxRetries)
+}
+
+func processFirehoseBlocks(
+	endpoint string,
+	apiToken string,
+	startBlock *int,
+	endBlock uint64,
+	batchSize int,
+	workerCount int,
+	bufferSize int,
+	chainID *big.Int,
+	externalRpc string,
+	handler func(blocks []*types.Block, batchNum int) error,
+) error {
+	client, closeFunc, grpcOpts, err := client.NewFirehoseClient(endpoint, apiToken, "", false, false)
+	if err != nil {
+		return fmt.Errorf("failed to create Firehose client: %w", err)
+	}
+	defer closeFunc()
+	grpcOpts = append(grpcOpts, grpc.UseCompressor(gzip.Name))
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
```

Comment:
Why this 5 minutes context timeout? I think this forces a stream kill each 5 minutes ...